csharp/src/Ice/ObjectAdapter.cs

```diff
@@ -1193,6 +1068,132 @@ private void UpdateLocatorRegistry(LocatorInfo? locatorInfo, IObjectPrx? proxy)
+            lock (_mutex)
+            {
+                Debug.Assert(_state == State.Uninitialized);
+                // Update the locator registry. We set state to State.Activating to prevent deactivation from
+                // other threads while the call is performed.
+                _state = State.Activating;
+            }
+
+            try
+            {
+                await UpdateLocatorRegistryAsync(_locatorInfo,
+                                                 CreateDirectProxy(new Identity("dummy", ""),
+                                                 IObjectPrx.Factory));
+            }
+            catch
+            {
+                // If we couldn't update the locator registry, we let the exception go through and don't activate the
+                // adapter to allow to user code to retry activating the adapter later.
+                lock (_mutex)
+                {
+                    _state = State.Uninitialized;
+                }
+                throw;
+            }
+
+            if ((Communicator.GetPropertyAsBool("Ice.PrintAdapterReady") ?? false) && Name.Length > 0)
+            {
+                Console.Out.WriteLine($"{Name} ready");
+            }
+
+            lock (_mutex)
+            {
+                Debug.Assert(_state == State.Activating);
+                foreach (IncomingConnectionFactory factory in _incomingConnectionFactories)
+                {
+                    factory.Activate();
+                }
+                _state = State.Active;
+            }
+        }
+
+        private async Task PerformDeactivateAsync()
+        {
+            // Wait for activation to complete. This is necessary avoid out of order locator updates.
+            if (_activateTask != null)
+            {
+                await _activateTask.ConfigureAwait(false);
+            }
+
+            lock (_mutex)
+            {
+                Debug.Assert(_state <= State.Deactivating);
+                _state = State.Deactivating;
+            }
+
+            // Note: the router/locator infos and incoming connection factory list are immutable at this point.
+            try
+            {
+                if (_routerInfo != null)
+                {
+                    // Remove entry from the router manager.
+                    Communicator.EraseRouterInfo(_routerInfo.Router);
+
+                    // Clear this object adapter with the router.
+                    _routerInfo.Adapter = null;
+                }
+
+                await UpdateLocatorRegistryAsync(_locatorInfo, null);
+            }
+            catch
+            {
+                // We can't throw exceptions in deactivate so we ignore failures to update the locator registry.
+            }
+
+            lock (_mutex)
+            {
+                if (_directCount > 0)
+                {
+                    Debug.Assert(_directCountSemaphore == null);
+                    _directCountSemaphore = new SemaphoreSlim(_directCount);
```

Comment:
You're allowing `_directCount` threads to enter the semaphore here... the `WaitAsync` below will therefore not wait. I think it should be `SemaphoreSlim(0)` and instead release the semaphore when `_directCount == 0`